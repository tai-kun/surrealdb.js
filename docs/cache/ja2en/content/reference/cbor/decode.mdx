---
title: decode
---

## Synchronous Decoding

The `decode` function decodes a CBOR-formatted `Uint8Array` into a JavaScript value.

### Import

```ts
import { decode } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decode(input: Uint8Array, options?: DecodeOptions): unknown;
```

#### Arguments

##### `input`

The CBOR-formatted `Uint8Array` to decode into a JavaScript value.

##### `options`

Decoding options.

`mapType?: "Object" | "Map"`
:   This option can be used to choose whether to decode key-value objects as JavaScript plain objects or `Map` objects. Choose `"Object"` to decode as a plain object and `"Map"` to decode as a `Map` class instance. The default is `"Object"`. Plain object keys can only be strings or numbers (i.e., `string | number`), but if you decode with this option set to `"Map"`, the keys can be any decoded value.

`reviver?: Reviver | ReviverObject | (Reviver | ReviverObject)[]`
:   This option is used to transform CBOR simple data items or tagged data items into JavaScript values. Simple data items are limited to values from 0 to 19 and 32 to 255. The tag of a tagged data item is a `number` or `bigint` from 0 to 2^64-1. If the function returns the `CONTINUE` symbol, which can be obtained with `Symbol.for("@tai-kun/surrealdb/cbor/continue")`, subsequent functions will be called. If the value cannot be transformed, `SurrealTypeError` is thrown. 

`isSafeMapKey?: (key: unknown, map: Map<unknown, unknown>) => boolean`
:   This option is a function that validates whether a key of a `Map` object is safe. The function takes two arguments: `key` and `map`. `key` is the value of the key, and `map` is the `Map` object being validated. If this function returns `false`, `CborUnsafeMapKeyError` is thrown. The default considers `"__proto__"` or `"constructor"` as unsafe keys. Note that specifying this option will completely override the default validation. If you want to add your own validation to the default validation, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`isSafeObjectKey?: (key: unknown, obj: Record<string | number, unknown>) => boolean`
:   This option is a function that validates whether a key of a plain object is safe. The function takes two arguments: `key` and `obj`. `key` is the value of the key, and `obj` is the plain object being validated. If this function returns `false`, `CborUnsafeMapKeyError` is thrown. The default considers `"__proto__"` or `"constructor"` as unsafe keys. Note that specifying this option will completely override the default validation. If you want to add your own validation to the default validation, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`maxDepth?: number`
:   Specifies the maximum depth of the object. The depth is incremented by 1 for each object or array entered. The default value is `64`. Since recursion is not used for decoding, there is no upper limit to the depth, but it is recommended to keep it small.

#### Return Value

The JavaScript value decoded from the CBOR-formatted `Uint8Array`.

### Example

#### Decoding as a `Map` object

```ts
import { decode } from "@tai-kun/surrealdb/cbor";

const cbor = new Uint8Array([
  0xa1, // mt: 5, ai: 1
  // key
  0x00, // mt: 0, ai: 0
  // value
  0x00, // mt: 0, ai: 0
]);

const value = decode(bytes, { mapType: "Map" });

console.log(value); // Map(1) { [ 0 ] => 0 }
```

## Asynchronous Decoding

The `decodeStream` function decodes a CBOR-formatted `Uint8Array` into a JavaScript value.

### Import

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decodeStream(
  input: ReadableStream<Uint8Array>,
  options?: DecodeStreamOptions,
): unknown;
```

#### Arguments

##### `input`

A readable stream of `Uint8Array`. The Fetch API may provide this stream on the `body` property of the `Response` object in response to a request.

##### `options`

In addition to the options for synchronous decoding, the following options can be specified:

`signal?: AbortSignal`
:   An AbortSignal to abort the stream.

#### Return Value

A `StatefulPromise` that resolves with the JavaScript value.

### Example

#### Fetch API

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";

const response = await fetch("http://localhost:8000/rpc", {
  method: "POST",
  headers: {
    Accept: "application/cbor",
    // ... more headers
  },
  // ... more properties
})

const result = await decodeStream(resp.body);
```
