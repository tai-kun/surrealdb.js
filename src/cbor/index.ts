export type * from "./decode-stream";
export { default as decodeStream } from "./decode-stream";

export type * from "./decode";
export { default as decode } from "./decode";

export {
  CONTINUE,
  type IsSafeMapKey,
  type IsSafeObjectKey,
  type Reviver,
  type ReviverObject,
  type SimpleItemReviver,
  type TaggedItemReviver,
} from "./decoder";

export type * from "./encode";
export { default as encode } from "./encode";

export {
  type AdditionalInfo,
  AI_EIGHT_BYTES,
  AI_FOUR_BYTES,
  AI_INDEFINITE_NUM_BYTES,
  AI_ONE_BYTE,
  AI_SIMPLE_FALSE,
  AI_SIMPLE_NULL,
  AI_SIMPLE_TRUE,
  AI_SIMPLE_UNDEFINED,
  AI_TWO_BYTES,
  type DataItem,
  type DataItemValue,
  HEADER_FALSE,
  HEADER_FLOAT_DOUBLE,
  HEADER_FLOAT_HALF,
  HEADER_NULL,
  HEADER_TRUE,
  HEADER_UNDEFINED,
  type MajorType,
  MT_ARRAY,
  MT_BYTE_STRING,
  MT_MAP,
  MT_NEGATIVE_INTEGER,
  MT_SIMPLE_FLOAT,
  MT_TAG,
  MT_UNSIGNED_INTEGER,
  MT_UTF8_STRING,
  Simple,
} from "./spec";

export type * from "./tagged";
export { default as Tagged } from "./tagged";

export type { ToCBOR } from "./traits";
export { canToCBOR } from "./traits";

export {
  type Replacer,
  write,
  writeBigInt,
  writeBoolean,
  writeByteString,
  writeEncodedUtf8String,
  writeHeader,
  writeNullable,
  writeNumber,
  writePayload,
  writeUtf8String,
} from "./write-utils";

export type * from "./writer"; // `Writer` も型だけでよし
