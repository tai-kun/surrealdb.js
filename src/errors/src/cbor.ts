import { SurrealError, type SurrealErrorOptions } from "./general";

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborerror)
 */
export class CborError extends SurrealError {
  static {
    this.prototype.name = "CborError";
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborwellformednesserror)
 */
export class CborWellFormednessError extends CborError {
  static {
    this.prototype.name = "CborWellFormednessError";
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.2
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbortoomuchdataerror)
 */
export class CborTooMuchDataError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborTooMuchDataError";
  }

  constructor(options?: SurrealErrorOptions | undefined) {
    super("Unconsumed input bytes remain after decoding.", options);
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.4
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbortoolittledataerror)
 */
export class CborTooLittleDataError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborTooLittleDataError";
  }

  constructor(options?: SurrealErrorOptions | undefined) {
    super(
      "Input data appears truncated or incomplete for CBOR decoding.",
      options,
    );
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.6
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborsyntaxerror)
 */
export class CborSyntaxError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborSyntaxError";
  }
}

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbormaxdepthreachederror)
 */
export class CborMaxDepthReachedError extends CborError {
  static {
    this.prototype.name = "CborMaxDepthReachedError";
  }

  constructor(
    public maxDepth: number,
    options?: SurrealErrorOptions | undefined,
  ) {
    super(`Maximum depth of ${maxDepth} has been reached.`, options);
  }
}

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborunsafemapkeyerror)
 */
export class CborUnsafeMapKeyError extends CborError {
  static {
    this.prototype.name = "CborUnsafeMapKeyError";
  }

  constructor(
    public key: unknown,
    options?: SurrealErrorOptions | undefined,
  ) {
    super(
      `Invalid key for CBOR map: ${String(key)}. The key must be a valid CBOR`
        + " data type that is safe to use in a JavaScript map or object.",
      options,
    );
  }
}

export class CborDecodeStreamAbortFailedError extends CborError {
  static {
    this.prototype.name = "CborStreamAbortFailedError";
  }

  override cause: unknown[];

  constructor(
    errors: readonly unknown[],
    options?: Omit<SurrealErrorOptions, "cause"> | undefined,
  ) {
    super("Failed to abort decode-stream", options);
    this.cause = errors.slice();
  }
}

export class CborMemoryError extends CborError {
  static {
    this.prototype.name = "CborMemoryError";
  }
}

export class CborMemoryBlockError extends CborError {
  static {
    this.prototype.name = "CborMemoryBlockError";
  }
}

type MemoryAddress = string | number;

export class CborMemoryBlockConflictError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborMemoryBlockConflictError";
  }

  constructor(
    address: MemoryAddress,
    expectedSize: number,
    actualSize: number,
  ) {
    super(
      `Memory conflict at address ${address}: `
        + `Expected size ${expectedSize}, but found ${actualSize}.`,
    );
  }
}

export class CborMemoryBlockInUseError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborMemoryBlockInUseError";
  }
}

export class CborUndefinedMemoryBlockError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborUndefinedMemoryBlockError";
  }

  constructor(address: MemoryAddress) {
    super(`Memory block at address ${address} is not defined.`);
  }
}
