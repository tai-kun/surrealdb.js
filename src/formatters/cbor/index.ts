export type * from "./formatter";
export { default } from "./formatter";
