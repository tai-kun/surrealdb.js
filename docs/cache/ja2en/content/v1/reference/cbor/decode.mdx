---
title: Decode
slug: en/v1/reference/cbor/decode
---

## Synchronous Decoding

The `decode` function decodes a `Uint8Array` in CBOR format into a JavaScript value.

### Import

```ts
import { decode } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decode(input: Uint8Array, options?: DecodeOptions): unknown;
```

#### Arguments

##### `input`

A `Uint8Array` in CBOR format to be decoded into a JavaScript value.

##### `options`

Decoding options.

`mapType?: "Object" | "Map"`
:   This option can be used to choose whether to decode key-value objects as JavaScript plain objects or `Map` objects. Selecting `"Object"` decodes as a plain object, while `"Map"` decodes as an instance of the `Map` class. The default value is `"Object"`. Plain object keys can only accept strings or numbers (i.e. `string | number`), but decoding with this option set to `"Map"` allows keys to accept all decoded values.

`reviver?: Reviver | ReviverObject | (Reviver | ReviverObject)[]`
:   This option is used to convert simple or tagged data items from CBOR to JavaScript values. Simple data items are limited to values between 0 and 19 and between 32 and 255. The tag of a tagged data item is a `number` or `bigint` between 0 and 2^64-1. If the function returns the `CONTINUE` symbol obtained with `Symbol.for("@tai-kun/surrealdb/cbor/continue")`, subsequent functions are called. If no conversion could be made, a `SurrealTypeError` is thrown.

`isSafeMapKey?: (key: unknown, map: Map<unknown, unknown>) => boolean`
:   This option is a function that verifies whether a key in a `Map` object is safe. The function accepts two arguments: `key`, which is the key value, and `map`, which is the `Map` object being validated. If this function returns `false`, a `CborUnsafeMapKeyError` is thrown. By default, it determines `"__proto__"` or `"constructor"` as unsafe keys. Please note that specifying this option completely overrides the default validation. If you want to add your own validation to the default validation, you should also determine `"__proto__"` or `"constructor"` as unsafe keys.

`isSafeObjectKey?: (key: unknown, obj: Record<string | number, unknown>) => boolean`
:   This option is a function that verifies whether a key in a plain object is safe. The function accepts two arguments: `key`, which is the key value, and `obj`, which is the plain object being validated. If this function returns `false`, a `CborUnsafeMapKeyError` is thrown. By default, it determines `"__proto__"` or `"constructor"` as unsafe keys. Please note that specifying this option completely overrides the default validation. If you want to add your own validation to the default validation, you should also determine `"__proto__"` or `"constructor"` as unsafe keys.

`maxDepth?: number`
:   Specifies the maximum depth of objects. The depth increases by 1 for every object or array entry. The default value is `64`. While decoding does not use recursive processing, there is no upper limit on depth, but it is recommended that it not be too large.

#### Return value

The JavaScript value decoded from the `Uint8Array` in CBOR format.

### Example

#### Decoding with `Map` objects

```ts
import { decode } from "@tai-kun/surrealdb/cbor";

const cbor = new Uint8Array([
  0xa1, // mt: 5, ai: 1
  // key
  0x00, // mt: 0, ai: 0
  // value
  0x00, // mt: 0, ai: 0
]);

const value = decode(bytes, { mapType: "Map" });

console.log(value); // Map(1) { [ 0 ] => 0 }
```

## Asynchronous Decoding

The `decodeStream` function decodes a `Uint8Array` in CBOR format into a JavaScript value.

### Import

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decodeStream(
  input: ReadableStream<Uint8Array>,
  options?: DecodeStreamOptions,
): unknown;
```

#### Arguments

##### `input`

A readable stream of `Uint8Array`. The Fetch API may provide this stream in the `body` property of the `Response` object for requests.

##### `options`

In addition to the options for synchronous decoding, the following options can be specified:

`signal?: AbortSignal`
:   An abort signal to interrupt the stream.

#### Return value

A `StatefulPromise` that resolves to a JavaScript value.

### Example

#### Fetch API

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";

const response = await fetch("http://localhost:8000/rpc", {
  method: "POST",
  headers: {
    Accept: "application/cbor",
    // ... more headers
  },
  // ... more properties
})

const result = await decodeStream(resp.body);
```
