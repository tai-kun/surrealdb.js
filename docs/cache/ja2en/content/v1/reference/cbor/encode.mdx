---
title: encode
---

The `encode` function encodes a JavaScript value into a CBOR-formatted `Uint8Array`.

### Import

```ts
import { encode } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function encode(value: unknown, options?: EncodeOptions): Uint8Array;
```

#### Arguments

##### `value`

The JavaScript value to be encoded into a CBOR-formatted `Uint8Array`. The following JavaScript values can be passed to `value`:

- Primitive values
  - `number`
  - `bigint`
  - `string`
  - `boolean`
  - `null`
  - `undefined`
- Objects with a `.toCBOR()` method
- Arrays or `Set`
- Plain objects or `Map`
- `Uint8Array`
- `Simple` objects
- Values that can be replaced with an optional `replacer`

If the `value` is not a primitive value and has a `toCBOR` property that is a function, the function will be called preferentially.

##### `options`

Encoding options.

`replacer?: Replacer | Replacer[]`
:   The encoder encodes most values to CBOR by default, but special values that do not fit this pattern can be replaced with encodable values. This option can be used to implement the replacement process by specifying a function or an array of functions. If the function returns the `CONTINUE` symbol obtained with `Symbol.for("@tai-kun/surrealdb/cbor/continue")`, subsequent functions will be called. If no replacement is possible, a `SurrealTypeError` will be thrown.

`isSafeMapKey?: (key: unknown, map: ReadonlyMap<unknown, unknown>) => boolean`
:   This option is a function that verifies whether the keys of a `Map` object are safe. The function accepts two arguments: `key` and `map`. `key` is the value of the key, and `map` is the `Map` object being validated. If this function returns `false`, a `CborUnsafeMapKeyError` will be thrown. By default, `"__proto__"` or `"constructor"` are considered unsafe keys. Note that specifying this option will completely override the default validation. If you want to add your own validation to the default validation, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`isSafeObjectKey?: (key: unknown, obj: { readonly [p: string]: unknown }) => boolean`
:   This option is a function that verifies whether the keys of a plain object are safe. The function accepts two arguments: `key` and `obj`. `key` is the value of the key, and `obj` is the plain object being validated. If this function returns `false`, a `CborUnsafeMapKeyError` will be thrown. By default, `"__proto__"` or `"constructor"` are considered unsafe keys. Note that specifying this option will completely override the default validation. If you want to add your own validation to the default validation, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`maxDepth?: number`
:   Specifies the maximum depth of objects. The depth increases by 1 for each entry into an object or array. The default value is `64`. Encoding does not use recursion, so there is no upper limit on the depth, but it is recommended not to make it too large.

#### Return Value

A `Uint8Array` containing the JavaScript value encoded into CBOR format.

### Examples

#### Protecting Object Keys with `isSafeObjectKey`

```ts
import { encode } from "@tai-kun/surrealdb/cbor";

encode(
  {
    "API_KEY": "***"
  },
  {
    isSafeObjectKey: k => !["constructor", "__proto__", "API_KEY"].includes(k),
  }
); // throws CborUnsafeMapKeyError
```

#### Replacing Symbols with `replacer`

```ts
import { CONTINUE, encode } from "@tai-kun/surrealdb/cbor";

const bytes = encode(Symbol.for("ID"), {
  replacer(o) {
    switch (o) {
      case Symbol.for("ID"):
        return "ID";

      default:
        return CONTINUE;
    }
  },
});

console.log(bytes); // Uint8Array(3) [ 98, 73, 68 ]
```

#### Limiting Maximum Depth with `maxDepth`

```ts
import { encode } from "@tai-kun/surrealdb/cbor";

encode(
  {
    // depth: 1
    a: new Map([
      // depth: 2
      ["b", [
        // depth: 3
        new Set([
          // depth: 4
        ]),
      ]],
    ]),
  },
  {
    maxDepth: 3,
  },
); // throws CborMaxDepthReachedError
```
