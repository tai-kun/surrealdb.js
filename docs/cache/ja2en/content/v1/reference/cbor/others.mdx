---
title: Other Tools
---

## write

### Import

```ts
import { write } from "@tai-kun/surrealdb/cbor";
```

### Syntax

`write` is a helper function that writes any JavaScript value that can be encoded into CBOR format to a `Writer`.

```ts
write(writer: Writer, value: unknown, options?: WriteOptions): void;
```

#### Arguments

##### `writer`

A `Writer` to write the value to as a `Uint8Array`.

##### `value`

A JavaScript value to be encoded to a CBOR formatted `Uint8Array`. The following JavaScript values can be passed to `value`:

- Primitive values
  - `number`
  - `bigint`
  - `string`
  - `boolean`
  - `null`
  - `undefined`
- Objects with `.toCBOR()` method.
- Array or `Set`.
- Plain object or `Map`.
- `Uint8Array`.
- `Simple` object.
- A value that can be replaced by an optional `replacer`.

If the value is not a primitive value and has a `toCBOR` property that is a function, it will be called preferentially.

##### `options`

Writing options.

`replacer?: Replacer | Replacer[]`
:   The encoder encodes most values to CBOR by default, but special values that do not fit may be replaced with encodable values. This option can be used to implement the replacement process by specifying a function or an array of functions. If the function returns the `CONTINUE` symbol obtained by `Symbol.for("@tai-kun/surrealdb/cbor/continue")`, subsequent functions will be called. If the value cannot be replaced, `SurrealTypeError` will be thrown.

`isSafeMapKey?: (key: unknown, map: ReadonlyMap<unknown, unknown>) => boolean`
:   This option is a function that verifies that the keys of a `Map` object are safe. The function takes two arguments, `key` and `map`. `key` is the value of the key, and `map` is the `Map` object being verified. If this function returns `false`, `CborUnsafeMapKeyError` will be thrown. By default, `"__proto__"` or `"constructor"` is considered an unsafe key. Note that specifying this option will completely overwrite the default verification. If you want to add your own verification to the default verification, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`isSafeObjectKey?: (key: unknown, obj: { readonly [p: string]: unknown }) => boolean`
:   This option is a function that verifies that the keys of a plain object are safe. The function takes two arguments, `key` and `obj`. `key` is the value of the key, and `obj` is the plain object being verified. If this function returns `false`, `CborUnsafeMapKeyError` will be thrown. By default, `"__proto__"` or `"constructor"` is considered an unsafe key. Note that specifying this option will completely overwrite the default verification. If you want to add your own verification to the default verification, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

#### Return value

None.

## writeNumber

### Import

```ts
import { writeNumber } from "@tai-kun/surrealdb/cbor";
```

### Syntax

`writeNumber` is a helper function that writes a number to a `Writer`.

```ts
writeNumber(writer: Writer, value: number): void;
```

#### Arguments

##### `writer`

A `Writer` to write the value to as a `Uint8Array`.

##### `value`

A number. For integers, values from -2^53-1 to 2^53-1 are valid. In JavaScript, there is no way to determine `.0` as a floating point number, so it will be encoded as an integer value to CBOR. For example, `-1.0` is `-1` and is an integer.

#### Return value

None.

## writeBigInt

### Import

```ts
import { writeInteger } from "@tai-kun/surrealdb/cbor";
```

### Syntax

`writeBigInt` is a helper function that writes a `bigint` to a `Writer`.

```ts
writeBigInt(writer: Writer, value: bigint): void;
```

#### Arguments

##### `writer`

A `Writer` to write the value to as a `Uint8Array`.

##### `value`

A `bigint`. Values from -2^64 to 2^64-1 are valid.

#### Return value

None.

## writeByteString

### Import

```ts
import { writeByteString } from "@tai-kun/surrealdb/cbor";
```

### Syntax

`writeByteString` is a helper function that writes a `Uint8Array` to a `Writer`.

```ts
writeByteString(writer: Writer, value: Uint8Array): void;
```

#### Arguments

##### `writer`

A `Writer` to write the value to as a `Uint8Array`.

##### `value`

A `Uint8Array`.

#### Return value

None.

## writeUtf8String

### Import

```ts
import { writeUtf8String } from "@tai-kun/surrealdb/cbor";
```

### Syntax

`writeUtf8String` is a helper function that writes a UTF-8 string to a `Writer`.

```ts
writeUtf8String(writer: Writer, value: string): void;
```

#### Arguments

##### `writer`

A `Writer` to write the value to as a `Uint8Array`.

##### `value`

A UTF-8 string.

#### Return value

None.

## writeEncodedUtf8String

### Import

```ts
import { writeEncodedUtf8String } from "@tai-kun/surrealdb/cbor";
```

### Syntax

`writeEncodedUtf8String` is a helper function that writes a UTF-8 encoded string (`Uint8Array`) to a `Writer`.

```ts
writeEncodedUtf8String(writer: Writer, value: Uint8Array): void;
```

#### Arguments

##### `writer`

A `Writer` to write the value to as a `Uint8Array`.

##### `value`

A UTF-8 encoded string (`Uint8Array`).

#### Return value

None.

## writeBoolean

### Import

```ts
import { writeBoolean } from "@tai-kun/surrealdb/cbor";
```

### Syntax

`writeBoolean` is a helper function that writes a boolean value to a `Writer`.

```ts
writeBoolean(writer: Writer, value: boolean): void;
```

#### Arguments

##### `writer`

A `Writer` to write the value to as a `Uint8Array`.

##### `value`

A boolean value.

#### Return value

None.

## writeNullable

### Import

```ts
import { writeNullable } from "@tai-kun/surrealdb/cbor";
```

### Syntax

`writeNullable` is a helper function that writes `null` or `undefined` to a `Writer`.

```ts
writeNullable(writer: Writer, value: null | undefined): void;
```

#### Arguments

##### `writer`

A `Writer` to write the value to as a `Uint8Array`.

##### `value`

`null` or `undefined`.

#### Return value

None.

## writeHeader

### Import

```ts
import { writeHeader } from "@tai-kun/surrealdb/cbor";
```

### Syntax

`writeHeader` is a helper function that writes the header of a CBOR data item to a `Writer`.

```ts
writeHeader(writer: Writer, mt: MajorType, length: number | bigint): void;
```

#### Arguments

##### `writer`

A `Writer` to write the value to as a `Uint8Array`.

##### `mt`

CBOR Major Type. This should be an integer value between 0 and 7.

##### `length`

The data itself or the length of the data in bytes.

#### Return value

None.

## writePayload

### Import

```ts
import { writePayload } from "@tai-kun/surrealdb/cbor";
```

### Syntax

`writePayload` is a helper function that writes the payload of a CBOR data item to a `Writer`.

```ts
writePayload(writer: Writer, value: Uint8Array): void;
```

#### Arguments

##### `writer`

A `Writer` to write the value to as a `Uint8Array`.

##### `value`

The payload of the CBOR data item.

#### Return value

None.
