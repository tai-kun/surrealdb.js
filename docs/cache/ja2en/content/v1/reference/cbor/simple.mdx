---
title: Simple
slug: en/v1/reference/cbor/simple
---

import Badge from "~/components/en/Badge.astro";

`Simple` is a simple value identified by an integer value from 0 to 255.

### Import

```ts
import { Simple } from "@tai-kun/surrealdb/cbor";
```

### `.constructor()`

Creates a new `Simple` object.

```ts
new Simple(value: number);
```

#### Arguments

##### `value`

An integer value from 0 to 255.

#### Return Value

When called via `new`, `Simple` returns its instance.

### `.value` <Badge variant="instance" /><Badge variant="property" />

The integer value from 0 to 255 that identifies the `Simple` object.

### Notes

Refer to the following materials:

- [https://datatracker.ietf.org/doc/html/rfc8949#name-simple-values](https://datatracker.ietf.org/doc/html/rfc8949#name-simple-values)
- [https://www.iana.org/assignments/cbor-simple-values/cbor-simple-values.xhtml](https://www.iana.org/assignments/cbor-simple-values/cbor-simple-values.xhtml)
