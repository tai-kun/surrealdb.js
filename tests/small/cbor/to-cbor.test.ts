import { encode, writeNumber, type Writer } from "@tai-kun/surrealdb/cbor";
import { expect, test } from "vitest";

const bytes = new Uint8Array([
  0b110_01010, // mt: 6, ai: 10
  0b100_00001, // mt: 4, ai:  1
  0b110_01011, // mt: 6, ai: 11
  0b000_10100, // mt: 0, ai: 20
]);

test("タグ付き", () => {
  const cborValue = {
    toCBOR: () => [10, [
      { toCBOR: () => [11, 20] },
    ]],
  };

  expect(encode(cborValue)).toStrictEqual(bytes);
});

test("タグなし", () => {
  const cborValue = {
    toCBOR: () => [20],
  };

  expect(encode(cborValue)).toStrictEqual(
    new Uint8Array([
      0b000_10100, // mt: 0, ai: 20
    ]),
  );
});

test("Writer で書き込む", () => {
  const cborValue = {
    toCBOR: (w: Writer) => {
      writeNumber(w, 20);
    },
  };

  expect(encode(cborValue)).toStrictEqual(
    new Uint8Array([
      0b000_10100, // mt: 0, ai: 20
    ]),
  );
});

test("3 つ以上の要素を持つ配列を返してエラー", () => {
  const cborValue = {
    toCBOR: () => [0, 1, 2],
  };

  expect(() => encode(cborValue)).toThrowErrorMatchingSnapshot();
});

test("空配列を返してエラー", () => {
  const cborValue = {
    toCBOR: () => [],
  };

  expect(() => encode(cborValue)).toThrowErrorMatchingSnapshot();
});
