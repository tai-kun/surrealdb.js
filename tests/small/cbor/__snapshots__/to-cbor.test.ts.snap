// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`3 つ以上の要素を持つ配列を返してエラー 1`] = `
{
  "actual": "an array of length 3",
  "expected": "an array of length 1 or 2",
  "message": "Expected an array of length 1 or 2 but got type String of value "an array of length 3".",
  "name": "SurrealValueError",
}
`;

exports[`空配列を返してエラー 1`] = `
{
  "actual": "an array of length 0",
  "expected": "an array of length 1 or 2",
  "message": "Expected an array of length 1 or 2 but got type String of value "an array of length 0".",
  "name": "SurrealValueError",
}
`;
